### PR TITLE
RPN loss fix for batch-by-default

### PR DESCRIPTION
**Pull request by armandmcqueen**
Moved RPN loss fix to sit on top of Can's new branch. Untested

### CODE REVIEW

**Comment by armandmcqueen** [comment]
Tested that it runs training steps without crashing

File: MaskRCNN_no_batch_convergence/model_box.py [hunk 1]
```diff
@@ -9,13 +9,38 @@
...
+
+
 @under_name_scope()
 def clip_boxes(boxes, window, name=None):
     """
     Args:
         boxes: nx4, xyxy
         window: [h, w]
     """
+    window = tf.squeeze(window, axis=0)
```
Comment (armandmcqueen) [inline MaskRCNN_no_batch_convergence/model_box.py:42]: @karakusc Is this an issue for batch-by-default?

Comment (karakusc) [reply MaskRCNN_no_batch_convergence/model_box.py:42]: I wrote `clip_boxes_batch` above this, we should be using that one.

File: MaskRCNN_no_batch_convergence/model_fpn.py [hunk 2]
```diff
@@ -216,8 +280,8 @@ def multilevel_rpn_losses(
                 name_scope='level{}'.format(lvl + 2))
             losses.extend([label_loss, box_loss])
 
-        total_label_loss = tf.add_n(losses[::2], name='label_loss')
-        total_box_loss = tf.add_n(losses[1::2], name='box_loss')
+        total_label_loss = tf.add_n(losses[::2], name='label_loss' if name else None)
```
Comment (vsuthichai) [inline MaskRCNN_no_batch_convergence/model_fpn.py]: Can probably get rid of these two lines now and have it back to the way it used to be.  Originally, It was to avoid the tensor naming issue for tensorboard.  It looks like the name_scope is being intentionally renamed within `multilevel_rpn_losses_batch_fixed_single_image` so it doesn't seem like it would appear to be a problem anymore.

File: MaskRCNN_no_batch_convergence/model_fpn.py [hunk 3]
```diff
@@ -401,6 +472,7 @@ def generate_fpn_proposals_batch_tf_op(multilevel_box_logits, multilevel_label_l
                 # print(anchor_list)
                 anchors = tf.stack(anchor_list)
 
+                """
```
Comment (karakusc) [inline MaskRCNN_no_batch_convergence/model_fpn.py:475]: Shouldnt' we be feeding `single_level_anchors_boxes` to `generate_bounding_box_proposals` below?

File: MaskRCNN_no_batch_convergence/model_fpn.py (hunk 2 above)
Comment (armandmcqueen) [reply MaskRCNN_no_batch_convergence/model_fpn.py]: Fixed

File: MaskRCNN_no_batch_convergence/model_fpn.py (hunk 3 above)
Comment (armandmcqueen) [reply MaskRCNN_no_batch_convergence/model_fpn.py:475]: Yes, we definitely should

Comment (armandmcqueen) [reply MaskRCNN_no_batch_convergence/model_fpn.py:475]: Fixed

Comment (karakusc) [reply MaskRCNN_no_batch_convergence/model_fpn.py:475]: I assume the convergence test had it the correct way?

**Comment by karakusc** [comment]
Looks good!

File: MaskRCNN_no_batch_convergence/model_fpn.py (hunk 3 above)
Comment (armandmcqueen) [reply MaskRCNN_no_batch_convergence/model_fpn.py:475]: Yeah. I had to fix merge conflicts when changing the PR to be on master and I think that's what introduced this problem. But I'll kick off another run as soon as its merged in just to make sure.

File: MaskRCNN_no_batch_convergence/train.py [hunk 4]
```diff
@@ -847,7 +834,7 @@ def do_predict(pred_func, input_file):
             callbacks.append(GPUUtilizationTracker())
 
         if args.perf:
-            callbacks.append(ThroughputTracker(args.images_per_step,
+            callbacks.append(ThroughputTracker(BATCH_SIZE_PLACEHOLDER*cfg.TRAIN.NUM_GPUS,
```
Comment (vsuthichai) [inline MaskRCNN_no_batch_convergence/train.py:837]: steps_per_epoch passed in as an argument to the training script isn't being accounted for here.  Causes the mean epoch throughput to be miscalculated.

Comment (armandmcqueen) [reply MaskRCNN_no_batch_convergence/train.py:837]: This is only an issue if we are deliberately passing in an incorrect steps_per_epoch values, right?

Looking at ThroughputTracker, I see 

`def __init__(self, items_per_step, items_per_epoch, trigger_every_n_steps=None, log_fn=None):`

so the code seems correct unless we are trying to have epochs be different than one full cycle of the data.

Comment (vsuthichai) [reply MaskRCNN_no_batch_convergence/train.py:837]: It's more to avoid accidentally not changing the steps_per_epoch at the training script level when changing batch sizes.  Could this be calculated internally?  Not a big deal either way.
